File: scripts/upload_validation.php
#!/usr/bin/env php
<?php

function check($filename) {

	$whitelist = ['ECHO', 'RETURN', 'PHP', 'NOP'];

	$out = [];
	$ret = 0;
	$cmd = [
		PHP_BINARY,
		"-d", "vld.active=1",
		"-d", "vld.execute=0",
		"-d", "extension=vld.so",
		"-d", "vld.format=1",
		"-d", "vld.col_sep=@",
		"-d", "log_errors=0",
		"-d", "error_log=/dev/null",
            	"-B", "if (!extension_loaded('vld')) die('no vld');",
		escapeshellarg($filename),
		'2>&1',
		];
	exec(implode(' ', $cmd), $out, $ret);
	if ($ret) {
        printf("Error: %d\n", $ret);
		return 2;
	}
	foreach($out as $line) {
		$sp = explode('@', $line);
		if (count($sp) < 5) {
			continue;
		}
		$opcode = $sp[4]; // # ,line, #, EIO, op, fetch, ext, return, operands
		if ($opcode && !in_array($opcode, $whitelist)) {
			printf("Upload_validation: Found an opcode: %s\n", $opcode);
			return 1;
		}
	}
	return 0;
}

if ($_SERVER['argc'] != 2) {
	die("Usage: {$_SERVER['argv']['0']} file_to_test.php\n");
}
exit(check($_SERVER['argv']['1']));

File: scripts/upload_validation.py
#!/usr/bin/python

import sys
import subprocess

WHITELIST = ('ECHO', 'RETURN', 'PHP', 'NOP')

def check(filename):
    try:
        output = subprocess.check_output(["php",
            "-d", "vld.active=1",
            "-d", "vld.execute=0",
            "-d", "extension=vld.so",
            "-d", "vld.format=1",
            "-d", "vld.col_sep=@",
            "-d", "log_errors=0",
            "-d", "error_log=/dev/null",
            "-B", "if (!extension_loaded('vld')) die('no vld');",
            filename],
            stderr=subprocess.STDOUT)
    except subprocess.CalledProcessError as e:
        print("Error: %s" % e)
        return 2

    for line in output.splitlines()[8:]:
        sp = line.split('@')
        if len(sp) < 5:
            continue
        opcode = sp[4]  # ,line, #, EIO, op, fetch, ext, return, operands
        if opcode not in WHITELIST:
            print("Upload_validation: Found an opcode: %s" % opcode)
            return 1
    return 0


if __name__ == '__main__':
    if len(sys.argv) != 2:
        print('Usage: %0 file_to_test.php', sys.argv[0])
    else:
        sys.exit(check(sys.argv[1]))
